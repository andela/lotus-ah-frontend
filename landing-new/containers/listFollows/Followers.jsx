// third-party libraries
import { connect } from 'react-redux';

// components
import Followers from '../../components/listFollows/Followers';

// actions
import markNotificationAsReadAction from '../../action/notification/readNotification';
import getFollowers from '../../action/follow/getFollowers';
import getFollowing from '../../action/follow/getFollowing';
import followUser from '../../action/follow/follow';

const mapStateToProps = state => ({
  auth: state.auth,
  notifications: state.getNotification,
  followers: state.follow.followers,
  following: state.follow.following,
  follow: state.follow
});

const mapDispatchToProps = {
  markNotificationAsRead: (token, id, mark) => markNotificationAsReadAction(token, id, mark),
  listFollows: userId => getFollowers(userId),
  listFollowing: userId => getFollowing(userId),
  followRequest: (userId, action) => followUser(userId, action),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Followers);
