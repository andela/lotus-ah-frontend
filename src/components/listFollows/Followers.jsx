// react libraries
import React, { Fragment, Component } from 'react';

// third-party libraries
import PropTypes from 'prop-types';

// components
import Header from '../reusables/header/Header';
import FollowDetail from './followDetail/FollowDetail';

/**
 * @class CreateArticle
 * @param {object} prevProps
 * @extends {Component}
 */
class Followers extends Component {
  state = {
    action: 'unfollow',
    text: 'Following',
    currentUserFollowId: null,
  };

  componentDidMount = () => {
    const {
      match: {
        params: { username },
      },
    } = this.props;
    const userId = username.split('_')[1];
    this.setState({
      targetUserId: userId,
    });
    this.props.listFollows(userId);
    this.props.listFollowing(userId);
  };

  componentDidUpdate = (prevProps) => {
    const { progress } = this.props.follow;
    const { action, text } = this.state;
    const newAction = action !== 'follow' ? 'follow' : 'unfollow';
    const newText = text !== 'Follow' ? 'Follow' : 'Following';
    if (progress === 'done' && prevProps.follow.progress === 'ongoing') {
      this.setState({
        action: newAction,
        text: newText,
      });
    }
  };

  /**
   * @param {object} event
   * @desc handles follow and unfollow button click
   * @memberof Followers
   */
  handleFollow = (event) => {
    const { id, action } = event.target.dataset;
    this.setState({
      currentUserFollowId: id,
    });
    this.props.followRequest(id, action)
      .then(() => {
        this.props.listFollows(this.state.targetUserId);
        this.props.listFollowing(this.state.targetUserId);
      });
  };

  render() {
    const user = JSON.parse(localStorage.getItem('user'));
    const { isAuth } = this.props.auth;
    const { followers, followersCount } = this.props.followers;
    const { following } = this.props.following;
    let allFollowers = [];
    if (followersCount > 0) {
      allFollowers = followers.map((userFollowing, index) => {
        let text = 'Follow';
        let action = 'follow';
        following.map((myFollowing) => {
          if (userFollowing.followerId === myFollowing.id) {
            text = 'Following';
            action = 'unfollow';
          }
        });
        return (
          <div className='list-item' key={index}>
            <FollowDetail
              userFollowing={userFollowing.User}
              userId={userFollowing.User.id}
              onClick={this.handleFollow}
              isAuth={isAuth}
              text={
                this.state.currentUserFollowId === userFollowing.User.id ? this.state.text : text
              }
              action={
                this.state.currentUserFollowId === userFollowing.User.id
                  ? this.state.action
                  : action
              }
            />
          </div>
        );
      });
    }
    return (
      <Fragment>
        <Header user={user} isAuth={isAuth} notifications={this.props.notifications} />
        <div className='list-following'>
          <div className='title'>
            <h4>List Followers</h4>
          </div>
          <div className='row d-flex flex-row items-container'>{allFollowers}</div>
        </div>
      </Fragment>
    );
  }
}
Followers.propTypes = {
  notifications: PropTypes.object,
  auth: PropTypes.object,
  listFollows: PropTypes.func,
  followers: PropTypes.object,
  listFollowing: PropTypes.func,
  following: PropTypes.object,
  followRequest: PropTypes.func,
  follow: PropTypes.object,
  match: PropTypes.object,
};

export default Followers;
