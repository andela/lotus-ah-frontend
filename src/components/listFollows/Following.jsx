// react libraries
import React, { Fragment, Component } from 'react';

// third-party libraries
import PropTypes from 'prop-types';

// components
import Header from '../reusables/header/Header';
import FollowDetail from './followDetail/FollowDetail';

/**
 * @class CreateArticle
 * @extends {Component}
 * @param {object} event
 */
class Following extends Component {
  state = {
    action: 'unfollow',
    text: 'Following',
    currentUserFollowId: null,
  };

  componentDidMount = () => {
    const {
      match: {
        params: { username },
      },
    } = this.props;
    const userId = username.split('_')[1];
    this.props.listFollowing(userId);
    this.setState({
      targetUserId: userId
    });
  };

  /**
   *@param {object} prevProps
   * @memberof Following
   */
  componentDidUpdate = (prevProps) => {
    const { progress } = this.props.follow;
    const { action, text } = this.state;
    const newAction = action !== 'follow' ? 'follow' : 'unfollow';
    const newText = text !== 'Follow' ? 'Follow' : 'Following';
    if (progress === 'done' && prevProps.follow.progress === 'ongoing') {
      this.setState({
        action: newAction,
        text: newText,
      });
    }
  };

  handleFollow = (event) => {
    const { id, action } = event.target.dataset;
    this.setState({
      currentUserFollowId: id,
    });
    this.props.followRequest(id, action);
  };

  render() {
    const user = JSON.parse(localStorage.getItem('user'));
    const { isAuth } = this.props.auth;
    const { following, followingCount } = this.props.following;
    let allFollowing = [];
    if (followingCount > 0) {
      allFollowing = following.map((userFollowing, index) => (
        <div className='list-item' key={index}>
          <FollowDetail
            userFollowing={userFollowing}
            userId={userFollowing.id}
            text={this.state.currentUserFollowId == userFollowing.id ? this.state.text : 'Following'}
            onClick={this.handleFollow}
            action={
              this.state.currentUserFollowId == userFollowing.id ? this.state.action : 'unfollow'
            }
            isAuth={isAuth}
          />
        </div>
      ));
    }
    return (
      <Fragment>
        <Header user={user} isAuth={isAuth} notifications={this.props.notifications} />
        <div className='list-following'>
          <div className='title'>
            <h4>List Following</h4>
          </div>
          <div className='row d-flex flex-row items-container'>{allFollowing}</div>
        </div>
      </Fragment>
    );
  }
}

Following.propTypes = {
  notifications: PropTypes.object,
  auth: PropTypes.object,
  listFollowing: PropTypes.func,
  following: PropTypes.object,
  follow: PropTypes.object,
  followRequest: PropTypes.func,
  match: PropTypes.object,
};

export default Following;
