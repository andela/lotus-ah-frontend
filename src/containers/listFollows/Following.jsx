// third-party libraries
import { connect } from 'react-redux';

// components
import Following from '../../components/listFollows/Following';

// actions
import markNotificationAsReadAction from '../../action/notification/readNotification';
import getFollowing from '../../action/follow/getFollowing';
import followUser from '../../action/follow/follow';

const mapStateToProps = state => ({
  auth: state.auth,
  notifications: state.getNotification,
  following: state.follow.following,
  follow: state.follow
});

const mapDispatchToProps = {
  markNotificationAsRead: (token, id, mark) => markNotificationAsReadAction(token, id, mark),
  listFollowing: userId => getFollowing(userId),
  followRequest: (userId, action) => followUser(userId, action),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Following);
