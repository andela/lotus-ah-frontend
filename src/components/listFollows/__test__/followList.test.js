// React Libraries
import React from 'react';

// third-party libraries
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';


// modules importation
import Followers from '../Followers';
import Following from '../Following';

configure({
  adapter: new Adapter(),
});

describe('Renders followers page', () => {
  const listFollows = () => {};
  const listFollowing = () => {};
  it('should render the followers component', () => {
    const component = shallow(
      <Followers
        auth={{
          isAuth: true
        }}
        followers={[]}
        following ={[]}
        listFollows ={
          listFollows
        }
        follow={{
          progress: 'done'
        }}
        listFollowing = {
          listFollowing
        }
        match={{
          params: {
            username: '@username_23'
          }
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });
});

describe('Renders following page', () => {
  const listFollowing = () => {};
  it('should render the following component', () => {
    const component = shallow(
      <Following
        auth={{
          isAuth: true,
        }}
        following={[]}
        listFollowing = {
          listFollowing
        }
        follow={{
          progress: 'done'
        }}
        match={{
          params: {
            username: '@username_23'
          }
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
