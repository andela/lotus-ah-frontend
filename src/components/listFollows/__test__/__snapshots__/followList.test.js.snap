// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders followers page should render the followers component 1`] = `
<Fragment>
  <Header
    isAuth={true}
    user={null}
  />
  <div
    className="list-following"
  >
    <div
      className="title"
    >
      <h4>
        List Followers
      </h4>
    </div>
    <div
      className="row d-flex flex-row items-container"
    />
  </div>
</Fragment>
`;

exports[`Renders following page should render the following component 1`] = `
<Fragment>
  <Header
    isAuth={true}
    user={null}
  />
  <div
    className="list-following"
  >
    <div
      className="title"
    >
      <h4>
        List Following
      </h4>
    </div>
    <div
      className="row d-flex flex-row items-container"
    />
  </div>
</Fragment>
`;
